README.md
python-random-vcard-generator
=============================

Generiert eine VCard mit zufälligen, aber plausiblen Daten.

https://de.wikipedia.org/wiki/VCard

Für die Erzeugung zufälliger Daten wird pyzufall benutzt.
https://github.com/davidak/pyzufall

Für die Erzeugung mehrerer VCards in einer Datei kann die Shell benutzt werden.
Das Funktioniert auf Unix (z.B. OS X) und Linux.

	for i in {1..5} ; do python3 vcard_generator.py >> Kontakte.vcf ; done

Beispiel
--------

	BEGIN:VCARD
	VERSION:3.0
	N:Reisner;Armin
	FN:Armin Reisner
	X-MAIDENNAME:Gerk
	BDAY;VALUE=DATE:1954-03-24
	CATEGORIES:Freunde
	EMAIL;TYPE=INTERNET:armin.reisner@web.de
	END:VCARD
	 
	BEGIN:VCARD
	VERSION:3.0
	N:Senn;Anica
	FN:Anica Senn
	CATEGORIES:Freunde
	EMAIL;TYPE=INTERNET:anica.senn@gmail.com
	URL;TYPE=WORK:http://anica-senn.net/
	NOTE:Motto: Auf Donner folgt gern Regen.
	END:VCARD
	 
	BEGIN:VCARD
	VERSION:3.0
	N:Johannes;Georg
	FN:Georg Johannes
	CATEGORIES:Familie
	EMAIL;TYPE=INTERNET:georg.johannes@web.de
	END:VCARD
	 
	BEGIN:VCARD
	VERSION:3.0
	N:Mucha;Sergej
	FN:Sergej Mucha
	CATEGORIES:Familie
	EMAIL;TYPE=INTERNET:sergej.mucha@hotmail.com
	NOTE:Motto: Betrunkene und Kinder sagen die Wahrheit.
	END:VCARD
	 
	BEGIN:VCARD
	VERSION:3.0
	N:Kehrer;Kaan
	FN:Kaan Kehrer
	X-MAIDENNAME:Pavlovic
	TITLE:Vermessungsingenieur
	CATEGORIES:Arbeit
	EMAIL;TYPE=INTERNET:kaan.kehrer@gmx.de
	NOTE:Motto: Alle Wege führen nach Rom.
	END:VCARD


CHANGELOG.rst
Changelog
=========

Version 0.9
-----------

Veröffentlicht am 05.06.2016

- Parameter -f ist nun optional, Ausgabe in Terminal möglich
- Weitere Encoding-Probleme mit Python 2.7 behoben
- Problem beim Speichern von mehr als 6.000.000 VCards behoben
- Performance enorm gesteigert (nur noch 4,7 statt 19 Sekunden für 10.000 VCards)
- RAM-Nutzung verringert auf 6 MB (vorher je nach Anzahl VCards, 100 MB bei 10.000)
- Prüfung, ob genug Speicherplatz verfügbar ist
- Spätere Dateigröße anzeigen

Version 0.8.3
-------------

Veröffentlicht am 23.05.2016

- Unicode Fehler behoben, Kompatibilität mit Python 2/3 verbessert
- Python 3.2 wird nicht mehr unterstützt

Version 0.8.2
-------------

Veröffentlicht am 07.07.2015

- Fehler bei Installation mit Python 3 behoben

Version 0.8.1
-------------

Veröffentlicht am 15.03.2015

- Fehler in Paket auf PyPI behoben

Version 0.8
-----------

Veröffentlicht am 15.03.2015

- läuft jetzt auch mit Python 2.7
- Geburtsname und Geburtsort hinzugefügt
- E-Mail, Nickname und Homepage hinzugefügt
- Kennung der Software, die die VCard erzeugt hat, hinzugefügt
- Motto entfernt, unrealistisch

Version 0.7
-----------

Veröffentlicht am 02.03.2014

- Fortschrittsanzeige mit vergangener und verbleibender Zeit
- Benutzung vereinfacht

Version 0.6
-----------

Veröffentlicht am 01.03.2014

- jetzt auch mit pip installierbar und auf PyPI gehostet
- Dokumentation entsprechend angepasst

Version 0.5
-----------

Veröffentlicht am 09.01.2014

- Kommandozeilen-Interface (CLI) hinzugefügt
- viele Verbesserungen am Code

Version 0.4
-----------

Veröffentlicht am 24.09.2013

- kompletter Rewrite. benutzt nun die Klasse Person von PyZufall
- PyZufall git Submodul gelöscht, da es nun über PyPI installierbar ist
- CHANGELOG hinzugefügt
- viele kleine Änderungen

Version 0.3
-----------

Veröffentlicht am 23.07.2013

- README und LICENSE hinzugefügt

Version 0.2
-----------

Veröffentlicht am 23.07.2013

- PyZufall als git Submodul eingebunden
- Docstring hinzugefügt

Version 0.1
-----------

Veröffentlicht am 23.07.2013

- einfacher VCard-Generator, der PyZufall benutzt


vcardgen.py
#!/usr/bin/env python3
# -*- coding: utf8 -*-
#
# Copyright (C) 2013 - 2014 davidak
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see {http://www.gnu.org/licenses/}.

import argparse
import random as r
from datetime import datetime as date
from pyzufall.person import Person
from .version import __version__

parser = argparse.ArgumentParser()
parser.add_argument('-V', '--version', action='version', version='Random VCard-Generator ' + __version__)
group = parser.add_mutually_exclusive_group()
group.add_argument("-v", "--verbose", action ="store_true", help="increase output verbosity")
group.add_argument("-q", "--quiet", action="store_true", help="no output")
parser.add_argument("-c", "--count", type=int, default=1, help="number of vcards to generate")
parser.add_argument("-o" ,"--output", default="Kontakte.vcf", help="output filename")
args = parser.parse_args()

gruppen = ['Arbeit', 'Kunden', 'Freunde', 'Familie', 'Sportverein', 'Ärzte', 'Piratenpartei', 'CCC', 'Bekannte aus dem Internet']

def generate_vcard():
	_p = Person()

	_s = "BEGIN:VCARD\n"
	_s += "VERSION:3.0\n"
	_s += "N:{};{};;;\n".format(_p.nachname, _p.vorname)
	_s += "FN:{}\n".format(_p.name)
	#_s+= "NICKNAME:Broho\n"
	#if r.randint(1,5) == 1:
		#_s += "X-MAIDENNAME:" + z.nachname() + "\n"
	_bday = date.strptime(_p.geburtsdatum, "%d.%m.%Y").date()
	_s += "BDAY;VALUE=DATE:{}\n".format(_bday)
	#_s = "ORG:" + z.firma() + ";Abteilung\n"
	if r.randint(1,4) == 1:
		_s += "TITLE:{}\n".format(_p.beruf)
	_s += "CATEGORIES:{}\n".format(r.choice(gruppen))
	#_s += "ADR;TYPE=WORK:;;Plorach Straße 27;Klostein;;46587;Deutschland\n"
	#_s += "EMAIL;TYPE=INTERNET:" + _p.email + "\n"
	#_s += "URL;TYPE=WORK:" + _p.webseite + "\n"

	# Notiz zusammensetzen
	_note = ''
	if r.randint(0,1):
		_note += 'Interessen: ' + _p.interessen
	if r.randint(0,1):
		if _note:
			_note += '\\n'
		_note += 'Lieblingsfarbe: ' + _p.lieblingsfarbe
	if r.randint(0,1):
		if _note:
			_note += '\\n'
		_note += 'Lieblingsessen: ' + _p.lieblingsessen
	if r.randint(0,1):
		if _note:
			_note += '\\n\\n'
		_note += 'Motto: ' + _p.motto
	if _note:
		_s += "NOTE:{}\n".format(_note)

	_s += "END:VCARD\n\n"
	return _s

def main():
	if not args.quiet:
		print("Random VCard-Generator {}\n".format(__version__))

	output = ''

	# Angegebene Anzahl an VCards generieren
	for i in range(args.count):
		output += generate_vcard()
		# wenn es länger dauert, Fortschritt anzeigen
		if args.count > 100:
			if args.verbose:
				print("VCard {} von {} generiert".format(i+1, args.count))
			if not args.verbose and not args.quiet:
				print('.', end='')

	# Zeilenumbruch nach Punkten
	if args.count > 100 and not args.verbose and not args.quiet:
		print()

	# VCards in Datei schreiben
	with open(args.output, 'w') as f:
		f.write(output)

	if not args.quiet:
		print("Fertig. {} VCard{} generiert und als {} gespeichert.".format(args.count, 's' if args.count > 1 else '', args.output))

if __name__ == "__main__":
	main()


vcardgen/main.py
#!/usr/bin/env python
# -*- coding: utf8 -*-
#
# Copyright (C) 2013 - 2016 davidak
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see {http://www.gnu.org/licenses/}.

from __future__ import (absolute_import, division, print_function, unicode_literals)
from builtins import *

from os import (statvfs, path)
import sys
import ctypes
import platform
import argparse
import frogress
import random as r
from datetime import datetime
from pyzufall.person import Person
from .version import __version__

if sys.version_info[0] == 2:
    from kitchen.text.converters import getwriter
    UTF8Writer = getwriter('utf8')
    sys.stdout = UTF8Writer(sys.stdout)

name = "Random VCard-Generator"

parser = argparse.ArgumentParser()
parser.add_argument('-V', '--version', action='version', version=name + ' ' + __version__)
parser.add_argument("-q", "--quiet", action="store_true", help="no output on screen")
parser.add_argument("-c", "--count", type=int, default=1, help="number of vcards to generate")
parser.add_argument('-f', '--file', help='filename or path, typical with .vcf extension')
args = parser.parse_args()

vcard_size = 312  # Bytes
total_vcard_size = vcard_size * args.count

widgets = [frogress.BarWidget, frogress.PercentageWidget, frogress.ProgressWidget('VCard: '), frogress.TimerWidget, frogress.EtaWidget]
gruppen = ['Arbeit', 'Kunden', 'Freunde', 'Familie', 'Sportverein', 'Ärzte', 'Piratenpartei', 'CCC', 'Bekannte aus dem Internet']


def sizeof_fmt(num, suffix='B'):
    """Dateigröße menschenlesbar ausgeben"""
    for unit in ['','Ki','Mi','Gi','Ti','Pi','Ei','Zi']:
        if abs(num) < 1024.0:
            return "%3.2f %s%s" % (num, unit, suffix)
        num /= 1024.0
    return "%.2f %s%s" % (num, 'Yi', suffix)


def get_free_space(path):
    """Gibt den freien Speicherplatz im angegebenen Verzeichnis in Bytes zurück"""
    if platform.system() == 'Windows':
        free_bytes = ctypes.c_ulonglong(0)
        ctypes.windll.kernel32.GetDiskFreeSpaceExW(ctypes.c_wchar_p(path), None, None, ctypes.pointer(free_bytes))
        return free_bytes.value
    else:
        st = statvfs(path)
        return st.f_bavail * st.f_frsize


def check_available_storage():
    free_space = get_free_space(path.dirname(path.realpath(args.file)))
    if total_vcard_size > free_space:
        print('Fehler: {} Speicherplatz benötigt, aber nur {} frei'.format(sizeof_fmt(total_vcard_size), sizeof_fmt(free_space)))
        sys.exit(1)


def generate_vcard():
    _p = Person()

    _s = "BEGIN:VCARD\n"
    _s += "VERSION:3.0\n"
    _s += "PRODID:-//davidak//{} {}//DE\n".format(name, __version__)
    _s += "FN:{}\n".format(_p.name)
    _s += "N:{};{};;;\n".format(_p.nachname, _p.vorname)
    if r.randint(0,1):
        _s+= "NICKNAME:{}\n".format(_p.nickname)
    if _p.geburtsname != _p.nachname:
        _s += "X-MAIDENNAME:{}\n".format(_p.geburtsname)
    _bday = datetime.strptime(_p.geburtsdatum, "%d.%m.%Y").date()
    _s += "BDAY:{}\n".format(_bday)
    if r.randint(0,1):
        _s += "BIRTHPLACE:{}\n".format(_p.geburtsort)
    #_s = "ORG:" + z.firma() + ";Abteilung\n"
    if _p.beruf != 'kein' and r.randint(0,1):
        _s += "TITLE:{}\n".format(_p.beruf)
    _s += "CATEGORIES:{}\n".format(r.choice(gruppen))
    _s += "TZ:+0100\n"
    #_s += "ADR;TYPE=WORK:;;Plorach Straße 27;Klostein;;46587;Deutschland\n"
    if _p.alter < 80 and r.randint(1,100) < 85:
        _s += "EMAIL;TYPE=INTERNET;TYPE=HOME;TYPE=PREF:{}\n".format(_p.email)
    if _p.alter < 50 and r.randint(0,1):
        _s += "URL;TYPE=HOME:{}\n".format(_p.homepage)

    # Notiz zusammensetzen
    _note = ''
    if r.randint(0,1):
        _note += 'Interessen: ' + _p.interessen
    if r.randint(0,1):
        if _note:
            _note += '\\n'
        _note += 'Lieblingsfarbe: ' + _p.lieblingsfarbe
    if r.randint(0,1):
        if _note:
            _note += '\\n'
        _note += 'Lieblingsessen: ' + _p.lieblingsessen
    if _note:
        _s += "NOTE:{}\n".format(_note)

    _s += "END:VCARD\n\n"

    del _p
    return _s


def main():
    if args.file:
        check_available_storage()
        output = ''
        if not args.quiet:
            print('{} {}'.format(name, __version__))
            print('Generiere {} ({})'.format(args.file, sizeof_fmt(total_vcard_size)))
            for i in frogress.bar(range(args.count), steps=args.count, widgets=widgets):
                output += generate_vcard()
                if (i % 250) == 0 or i == args.count - 1: # schreibe immer 250 vcards in datei
                    with open(args.file, mode='a', encoding='utf-8') as f:
                        f.write(output)
                        output = ''
            print("\n")
        else:
            for i in range(args.count):
                output += generate_vcard()
                if (i % 250) == 0 or i == args.count - 1: # schreibe immer 250 vcards in datei
                    with open(args.file, mode='a', encoding='utf-8') as f:
                        f.write(output)
                        output = ''
    else:
        for i in range(args.count):
            print(generate_vcard(), end='')


if __name__ == "__main__":
    main()


vcard_generator.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

"""
Der Python Random VCard-Generator generiert VCards mit zufälligen, aber plausiblen Daten.

Copyright (C) 2013 davidak

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see {http://www.gnu.org/licenses/}.
"""

import random as r
from datetime import datetime as date
from pyzufall.person import Person

__version__ = '0.4'

gruppen = ['Arbeit', 'Freunde', 'Familie', 'Sportverein', 'Ärzte', 'Piratenpartei', 'CCC', 'Bekannte aus dem Internet', 'Kegelclub', 'Mafia', 'Expartner', 'Prostituierte']


p = Person()

print("BEGIN:VCARD")
print("VERSION:3.0")
print("N:" + p.nachname + ";" + p.vorname + ";;;")
print("FN:" + p.name)
#print("NICKNAME:Broho")
#if r.randint(1,5) == 1:
#	print("X-MAIDENNAME:" + z.nachname())
bday = date.strptime(p.geburtsdatum, "%d.%m.%Y").date()
print("BDAY;VALUE=DATE:" + date.strftime(bday, "%Y-%m-%d"))
#print("ORG:" + z.firma() + ";Abteilung")
if r.randint(1,4) == 1: # jeder 5. Kontakt von Arbeit
	print("TITLE:" + p.beruf)
print("CATEGORIES:" + r.choice(gruppen))
#print("ADR;TYPE=WORK:;;Plorach Straße 27;Klostein;;46587;Deutschland")
#print("EMAIL;TYPE=INTERNET:" + p.email)
#print("URL;TYPE=WORK:" + p.webseite)
note = ''
if r.randint(0,1):
	note += 'Interessen: ' + p.interessen
if r.randint(0,1):
	if note:
		note += '\\n'
	note += 'Lieblingsfarbe: ' + p.lieblingsfarbe
if r.randint(0,1):
	if note:
		note += '\\n'
	note += 'Lieblingsessen: ' + p.lieblingsessen
if r.randint(0,1):
	if note:
		note += '\\n\\n'
	note += 'Motto: ' + p.motto
print("NOTE:" + note)
print("END:VCARD")
print('') #Zeilenumbruch
